src/model/idtypes.rs

```diff
@@ -0,0 +1,300 @@
+impl IdType for Album {
+    const TYPE: Type = Type::Album;
+}
+impl PlayContextIdType for Album {}
+impl IdType for Track {
+    const TYPE: Type = Type::Track;
+}
+impl PlayableIdType for Track {}
+impl IdType for Playlist {
+    const TYPE: Type = Type::Playlist;
+}
+impl PlayContextIdType for Playlist {}
+impl IdType for User {
+    const TYPE: Type = Type::User;
+}
+impl IdType for Show {
+    const TYPE: Type = Type::Show;
+}
+impl PlayContextIdType for Show {}
+impl IdType for Episode {
+    const TYPE: Type = Type::Episode;
+}
+impl PlayableIdType for Episode {}
+
+pub type ArtistId = Id<Artist>;
+pub type AlbumId = Id<Album>;
+pub type TrackId = Id<Track>;
+pub type PlaylistId = Id<Playlist>;
+pub type UserId = Id<User>;
+pub type ShowId = Id<Show>;
+pub type EpisodeId = Id<Episode>;
+
+pub type ArtistIdBuf = IdBuf<Artist>;
+pub type AlbumIdBuf = IdBuf<Album>;
+pub type TrackIdBuf = IdBuf<Track>;
+pub type PlaylistIdBuf = IdBuf<Playlist>;
+pub type UserIdBuf = IdBuf<User>;
+pub type ShowIdBuf = IdBuf<Show>;
+pub type EpisodeIdBuf = IdBuf<Episode>;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Artist {}
+impl private::Sealed for Artist {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Album {}
+impl private::Sealed for Album {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Track {}
+impl private::Sealed for Track {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Playlist {}
+impl private::Sealed for Playlist {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum User {}
+impl private::Sealed for User {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Show {}
+impl private::Sealed for Show {}
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Episode {}
+impl private::Sealed for Episode {}
+
+/// A Spotify object id of given [type](crate::model::enums::types::Type)
+///
+/// This is a not-owning type, it stores a &str only.
+/// See [IdBuf](crate::model::idtypes::IdBuf) for owned version of the type.
+#[derive(Debug, PartialEq, Eq, Serialize)]
+pub struct Id<T> {
+    #[serde(default)]
+    _type: PhantomData<T>,
+    #[serde(flatten)]
+    id: str,
+}
+
+/// A Spotify object id of given [type](crate::model::enums::types::Type)
+///
+/// This is an owning type, it stores a String.
+/// See [IdBuf](crate::model::idtypes::Id) for light-weight non-owning type.
+///
+/// Use `Id::from_id(val).to_owned()`, `Id::from_uri(val).to_owned()` or `Id::from_id_or_uri(val).to_owned()`
+/// to construct an instance of this type.
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+pub struct IdBuf<T> {
+    #[serde(default)]
+    _type: PhantomData<T>,
+    #[serde(flatten)]
+    id: String,
+}
+
+impl<T> AsRef<Id<T>> for IdBuf<T> {
+    fn as_ref(&self) -> &Id<T> {
+        // Safe, b/c of the same T between types, IdBuf can't be constructed from invalid id, and Id is just a wrapped str with ZST type tag
+        unsafe { &*(&*self.id as *const str as *const Id<T>) }
```

Comment:
If it's safe, is it fine to remove `unsafe` block ?

Reply:
The unsafe block doesn't mean the code is unsafe. It's needed here to let me do raw pointers type-casting and dereferencing. I can't do it without the `unsafe` block. It really means: “Hey, compiler, I know, you can't prove the code is safe, but I, a human programmer, do know better in this concrete time, I proved it to be safe, so keep calm, I take responsibility for all the type and memory checks”.

Reply:
Yes and it seems safe because `Id`'s only field is a `str`, and you seem to be doing `String -> str -> Id`. Doesn't this require `#[repr(C)]` though? Or I guess that's only for cases where there's padding between members? Also, why not just use a [transmute](https://doc.rust-lang.org/nomicon/transmutes.html)?

Reply:
I started with `transmute()` version, but then clippy told me I better use type-casting here. Also, transmute is an unsafe function as well, so using transmute doesn't give any benefits. `#[repr(C)]` here is unnecessary, because this struct has only one real field, which is a noop in Rust, so after compilation it has the same real memory layout as `str`.